### DISCUSSION TITLE
Clarification on Function Call Order During Invariant Testing with Foundry

### DISCUSSION THREAD

**Discussion by wildanvin** [Cyfrin/security-and-auditing-full-course-s23#132] [Q&A] [answered]
Hello, everyone,

I'm currently going throuh the Invariant part of the course and I got a question regarding the execution order of functions when using invariant testing. My question is about this [function](https://github.com/Cyfrin/sc-exploits-minimized/blob/7f7bc90560e22360f91330e6ef978b3892460c21/test/invariant-break/HandlerStatefulFuzz/Invariant.t.sol#L48)
Here's the code snippet of the function:

```solidity
    function statefulFuzz_testInvariantBreakHandler() public {
        vm.startPrank(owner);
 ->     handlerStatefulFuzzCatches.withdrawToken(mockUSDC);
 ->     handlerStatefulFuzzCatches.withdrawToken(yeildERC20);
        vm.stopPrank();

        assert(mockUSDC.balanceOf(address(handlerStatefulFuzzCatches)) == 0);
        assert(yeildERC20.balanceOf(address(handlerStatefulFuzzCatches)) == 0);
        assert(mockUSDC.balanceOf(owner) == startingAmount);
        assert(yeildERC20.balanceOf(owner) == startingAmount);
    }
```

My question revolves around the order in which `handlerStatefulFuzzCatches.withdrawToken(mockUSDC)` and `handlerStatefulFuzzCatches.withdrawToken(yeildERC20)` are executed in relation to the fuzzing process:

- Are these withdrawToken functions called at the beginning of the test, and then the fuzzer attempts to break the invariants by calling other functions?
- Or, are these specific withdrawToken calls supposed to be executed after the fuzzer has run through other functions?

Essentially, I'm trying to understand if the specified `withdrawToken` operations are intended to set up the state for invariant testing.

Thank you in advance for your help!

**Comment by EngrPips** [comment]
Hey @wildanvin , i could go on and on explaining the execution orders of functions when using invariant testing but honestly what would be of more value to you is to set the number of runs to 1 and run the test, That would make it so the fuzzer runs once and you should use the `-vvvv` flag also so it shows the stack trace of all of the executed transactions, That way you would see for yourself what are the execution orders of function.

**Reply by wildanvin** [reply, marked as answer]
Yes you are rigth, I forgot about the -vvvvv option in foundry... The withdrawToken functions are called at the beginning of the test, and then the fuzzer attempts to break the invariants by calling other functions specified in the handler. The process is basically as follows:

1. The `statefulFuzz_testInvariantBreakHandler` function is called (executing everything on it).
2. A random function from the Handler contract is called.
3. State is preserved.
4. Number 1 again until it breaks.

Basically, at step 1 we are withdrawing everything that the Handler (probably) deposited. Nice.

**Comment by EngrPips** [comment]
Awesome, I love that you see with your own eyes, instead of me sending just some text to explain. wish you the best on your knowledge journey
